Tensile/Components/MAC_F32.py

```diff
@@ -67,31 +72,86 @@ def __call__(self, writer, m, innerUnroll):
         priority = Component.Priority.find(writer)
         macIdx = 0
 
-        for iui in range(0, innerUnroll):
-            for idx1 in range(0, kernel["ThreadTile1"]):
-                for idx0 in range(0, kernel["ThreadTile0"]):
-                    vars["idx0"] = idx0
-                    vars["idx1"] = idx1
-                    vars["a"] = idx0 if writer.tPB["tile01Idx"] else idx1
-                    vars["b"] = idx1 if writer.tPB["tile01Idx"] else idx0
-                    vars["iui"] = iui
-
-                    vars["cStr"] = "v[vgprValuC + {idx0} + {idx1}*{ThreadTile0}]".format_map(vars)
-                    vars["aStr"] = "v[vgprValuA_X{m}_I{iui} + {a}]".format_map(vars)
-                    vars["bStr"] = "v[vgprValuB_X{m}_I{iui} + {b}]".format_map(vars)
-
-                    if instruction == "v_fma_f32":
-                        kStr += "v_fma_f32 {cStr}, {aStr}, {bStr}, {cStr}{endLine}".format_map(vars)
-                    else:
-                        kStr += "{instruction} {cStr}, {aStr}, {bStr}{endLine}".format_map(vars)
-
-                    kStr += priority(writer, 1, "Raise priority while processing macs")
-
-                    if macIdx == kernel["PerformanceWaitLocation"]:
-                        kStr += "s_waitcnt lgkmcnt({PerformanceWaitCount}) // extra wait for performance{endLine}".format_map(vars)
-                    if macIdx == kernel["PerformanceSyncLocation"]:
-                        kStr += "s_barrier // extra barrier for performance{endLine}".format_map(vars)
-                    macIdx += 1
+        if dualMacEnable == 1:
+            instQ = queue.Queue();
+            for iui in range(0, innerUnroll):
+               for idx1 in range(0, kernel["ThreadTile1"]):
+                   for idx0 in range(0, kernel["ThreadTile0"]):
+                        vars["idx0"] = idx0
+                        vars["idx1"] = idx1
+                        vars["a"] = idx0 if writer.tPB["tile01Idx"] else idx1
+                        vars["b"] = idx1 if writer.tPB["tile01Idx"] else idx0
+                        vars["iui"] = iui
+
+                        vars["cStr"] = "v[vgprValuC + {idx0} + {idx1}*{ThreadTile0}]".format_map(vars)
+                        vars["aStr"] = "v[vgprValuA_X{m}_I{iui} + {a}]".format_map(vars)
+                        vars["bStr"] = "v[vgprValuB_X{m}_I{iui} + {b}]".format_map(vars)
+
+                        instVars = {}
```

Comment:
Can we unify the for loops?
The difference is mac code generation part only.

We can have a if statement inside for loops (like this).

if dualMacEnable == 1:
    (dual mac generation part)
    instVars....
else:
    (existing mac generation)

We can always define instQ and keep it empty for dualMacEnable=0 case.
Then, We can unify PerformanceWaitLocation and PerformanceSyncLocation code.
if PerformanceSyncLocation:
   call pop all instructions function (does nothing for dualMacEnable=0)
   PerformanceWaitCount code

Reply:
Good suggestion. I will try to unify the code.